### PR TITLE
Transform more functions for HLSL

### PR DESCRIPTION
**Pull request by RobDangerous**
This adds everything requested in #128 and all of the remaining "various functions" from https://anteru.net/blog/2016/mapping-between-hlsl-and-glsl/

### CODE REVIEW

File: spirv_hlsl.cpp [hunk 1]
```diff
@@ -1268,22 +1339,217 @@ void CompilerHLSL::emit_instruction(const Instruction &instruction)
...
+		break;
+
+	case OpLogicalNot:
+	{
+		auto result_type = ops[0];
+		auto id = ops[1];
+		auto &type = get<SPIRType>(result_type);
+
+		if (type.vecsize > 1)
+			emit_unrolled_binary_op(result_type, id, ops[2], ops[3], "&&"); //**
```
Comment (HansKristian-ARM) [inline spirv_hlsl.cpp]: You'll need to add emit_unrolled_unary_op

Comment (HansKristian-ARM) [inline spirv_hlsl.cpp]: Also, should be "!", and not "&&" :)

Comment (RobDangerous) [inline spirv_hlsl.cpp]: Arrr ("//**" is my todo sign).

File: spirv_hlsl.cpp [hunk 2]
```diff
@@ -1213,33 +1242,73 @@ void CompilerHLSL::emit_uniform(const SPIRVariable &var)
 	statement(variable_decl(var), ";");
 }
 
+string CompilerHLSL::bitcast_glsl_op(const SPIRType &out_type, const SPIRType &in_type)
+{
```
Comment (HansKristian-ARM) [inline spirv_hlsl.cpp]: Tests needed for bitcasts

File: spirv_hlsl.cpp [hunk 3]
```diff
@@ -1213,33 +1242,73 @@ void CompilerHLSL::emit_uniform(const SPIRVariable &var)
...
+		return "asdouble";
+	else if (out_type.basetype == SPIRType::Double && in_type.basetype == SPIRType::UInt64)
+		return "asdouble";
+	else
+		return "";
+}
+
 void CompilerHLSL::emit_glsl_op(uint32_t result_type, uint32_t id, uint32_t eop, const uint32_t *args, uint32_t count)
 {
 	GLSLstd450 op = static_cast<GLSLstd450>(eop);
```
Comment (HansKristian-ARM) [inline spirv_hlsl.cpp]: Tests needed for GLSL builtins

File: spirv_hlsl.cpp [hunk 4]
```diff
@@ -1268,22 +1339,217 @@ void CompilerHLSL::emit_instruction(const Instruction &instruction)
...
+
+		if (type.vecsize > 1)
+			emit_unrolled_binary_op(result_type, id, ops[2], ops[3], "&&"); //**
+		else
+			UOP(!);
+		break;
+	}
+
+	case OpIEqual:
+	{
```
Comment (HansKristian-ARM) [inline spirv_hlsl.cpp]: Tests needed for vector variants here

File: spirv_hlsl.cpp [hunk 5]
```diff
@@ -608,9 +637,9 @@ void CompilerHLSL::emit_resources()
 
 		begin_scope();
 		sort(input_variables.begin(), input_variables.end(), variable_compare);
-		emit_builtin_inputs_in_struct();
```
Comment (HansKristian-ARM) [inline spirv_hlsl.cpp]: Any particular reason for moving this around?

Comment (RobDangerous) [inline spirv_hlsl.cpp]: Aye, Microsoft's compiler errored when TEXCOORDs appeared before an SV_SampleIndex.

Comment (HansKristian-ARM) [inline spirv_hlsl.cpp]: Fair enough.

File: spirv_glsl.cpp [hunk 6]
```diff
@@ -5016,6 +5039,22 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
 			require_extension("GL_OES_standard_derivatives");
 		break;
 
+	case OpDPdxFine:
+		UFOP(dFdxFine);
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Don't these functions require extensions?

Comment (RobDangerous) [inline spirv_glsl.cpp]: There's an extension for 4.0 (https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_derivative_control.txt) and it's core for 4.5. But nothing for ES I think.

Comment (RobDangerous) [inline spirv_glsl.cpp]: Do we error if the extension is unavailable?

Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Hm, Indeed, can't find any ES extension for this. Probably best to throw if ES, and require that extension if desktop GLSL < 4.5.

**Pull request merged**